src/test/java/io/confluent/examples/streams/CustomJoinWithTableTriggeringStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,429 @@
+      this.frequencyToCheckForExpiredWaitTimes = frequencyToCheckForExpiredWaitTimes;
+      this.streamBufferStoreName = streamBufferStoreName;
+      this.tableStoreName = tableStoreName;
+    }
+
+    @Override
+    public Transformer<String, Double, KeyValue<String, Pair<Double, Long>>> get() {
+      return new Transformer<String, Double, KeyValue<String, Pair<Double, Long>>>() {
+
+        private KeyValueStore<String, Pair<Double, Instant>> streamBufferStore;
+        private KeyValueStore<String, Long> tableStore;
+        private ProcessorContext context;
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+          streamBufferStore = (KeyValueStore<String, Pair<Double, Instant>>) context.getStateStore(streamBufferStoreName);
+          tableStore = (KeyValueStore<String, Long>) context.getStateStore(tableStoreName);
+          this.context = context;
+          // Note: In practice, you will probably want to use `PunctuationType.STREAM_TIME`.  However, using stream time
+          // would make the test/validation setup in this example more complicated, hence we use wall clock time.
+          this.context.schedule(frequencyToCheckForExpiredWaitTimes, PunctuationType.WALL_CLOCK_TIME, this::punctuate);
+        }
+
+        @Override
+        public KeyValue<String, Pair<Double, Long>> transform(final String key, final Double value) {
+          return sendFullJoinRecordOrWaitForTableSide(key, value);
+        }
+
+        private KeyValue<String, Pair<Double, Long>> sendFullJoinRecordOrWaitForTableSide(final String key, final Double value) {
+          final Long tableValue = tableStore.get(key);
+          if (tableValue != null) {
+            // We have data for both the stream and the table, so we can send a fully populated join message downstream
+            // immediately.
+            return KeyValue.pair(key, new Pair<>(value, tableValue));
+          } else {
+            // Don't send a join output just yet because we're still lacking table-side information.  Instead, buffer
+            // the current stream-side record, hoping that the table side will eventually see a matching record within
+            // `maxWaitTimePerRecordForTableSideData`.
+            streamBufferStore.put(key, new Pair<>(value, Instant.now()));
+            return null;
+          }
+        }
+
+        private void punctuate(final long timestamp) {
+          sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(timestamp);
+        }
+
+        private void sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(final long timestamp) {
+          try (KeyValueIterator<String, Pair<Double, Instant>> iterator = streamBufferStore.all()) {
+            while (iterator.hasNext()) {
+              final KeyValue<String, Pair<Double, Instant>> record = iterator.next();
+              final Duration delta = Duration.between(record.value.y, Instant.ofEpochMilli(timestamp));
```

Comment:
To get event-time semantics, we should compute the delta using `context.timestamp()` -- note that `context.timestamp()` returns current "stream-time" (because we don't have a record "at hand") that is exactly what we want for this purpose.

Reply:
Ack.  The example was originally (and intentionally) using wall clock time.  But apparently the feedback is that it should use event-time. ;-)  So it is now!

Reply:
The DSL is built on event-time, so if we argue "extend the DSL with a custom join" it should be event-time too, IMHO.

Of course, we can also add examples for processing time operators, but if we do, we need to be very explicit about the provided semantics.

Reply:
To be clear: the updated PR *is* already using event-time. :-)